src/utils/optionsdict.h

```diff
@@ -95,18 +95,26 @@ class OptionsDict : TypeDict<bool>,
   // is still in scope, when the parent pointer is used
   void AddSubdictFromString(const std::string& str);
 
+  // Returns the first option in the dict that has not been read to
+  // find syntax errors in options added using AddSubdictFromString
+  std::string FindNotUsed() const;
+
   bool HasSubdict(const std::string& name) const;
 
  private:
   const OptionsDict* parent_ = nullptr;
   std::map<std::string, OptionsDict> subdicts_;
+  mutable std::map<std::string, bool> used_;
```

Comment:
Maybe instead modify TypeDict?
```
template <typename T>
class TypeDict {
 protected:
  struct V {
      const T& Get() const { is_used_ = true; return value_; }
      void Set(const T& v) { is_used_ = false; value_ = v; }
      bool IsSet() const { return is_used_; }
    private:
      mutable bool is_used_ = false;
      T value_;
  };
  std::unordered_map<std::string, V> dict_;
};
```

Reply:
Is the added complexity worth it?

Reply:
It looks cleaner that way..
(no need to worry about dicts_ and used_ being in sync).
Also if there is a type mismatch (e.g. expected `gpu=2` but passed `gpu=2.0` instead), current code won't notice it.

Reply:
Also unsed empty subdictionary is not detected (with either approach), probably not a problem.

Reply:
Type mismatches were already caught, but now the error messages are more informative. Any suggestions for a cleaner way to do this?